processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java

```diff
@@ -71,6 +79,62 @@
   private final Supplier<GroupByQueryConfig> config;
   private final StupidPool<ByteBuffer> intermediateResultsBufferPool;
 
+  private static void getBytesFromBuffer(ByteBuffer src, byte[] dst, int srcOffset, int readLen)
```

Comment:
I remember now why I used absolute reads instead of relative: the relative reads change the position state of the ByteBuffer.

I made two versions of compare() and benchmarked them, one performs absolute reads and the other does relative reads on a read-only view of the ByteBuffer:
```
public int compareRelative(ByteBuffer o1, ByteBuffer o2)
      {
        int ret = 0;
        int dimIndex = 0;
        int maxDimIndex = dimHandlers.size();

        ByteBuffer o1ro = o1.asReadOnlyBuffer();
        ByteBuffer o2ro = o2.asReadOnlyBuffer();

        while (dimIndex < maxDimIndex) {
          DimensionHandler handler = dimHandlers.get(dimIndex);
          if (handler == null) {
            // Special time column handling
            int int1 = o1ro.getInt();
            int int2 = o2ro.getInt();
            ret = Integer.compare(int1, int2);
          } else {
            int valLen = handler.getEncodedValueSize();
            byte[] bytes1 = new byte[valLen];
            byte[] bytes2 = new byte[valLen];
            o1ro.get(bytes1, 0, valLen);
            o2ro.get(bytes2, 0, valLen);
            ret = handler.compareBytes(bytes1, bytes2);
          }

          if (ret != 0) {
            return ret;
          }

          dimIndex++;
        }
        return ret;
      }
```

```
public int compareAbsolute(ByteBuffer o1, ByteBuffer o2)
      {
        int pos = 0;
        int limit = o1.limit();
        int ret = 0;
        int dimIndex = 0;
        int maxDimIndex = dimHandlers.size();

        while (pos < limit && dimIndex < maxDimIndex) {
          DimensionHandler handler = dimHandlers.get(dimIndex);
          if (handler == null) {
            // Special time column handling
            int int1 = o1.getInt(pos);
            int int2 = o2.getInt(pos);
            pos += Ints.BYTES;
            ret = Integer.compare(int1, int2);
          } else {
            int valLen = handler.getEncodedValueSize();
            byte[] bytes1 = new byte[valLen];
            byte[] bytes2 = new byte[valLen];

            getBytesFromBuffer(o1, bytes1, pos, valLen);
            getBytesFromBuffer(o2, bytes2, pos, valLen);

            pos += valLen;

            ret = handler.compareBytes(bytes1, bytes2);
          }

          if (ret != 0) {
            return ret;
          }

          dimIndex++;
        }
        return ret;
      }
```

Based on benchmarking, the absolute read version is faster:
```
ABSOLUTE
GroupByBenchmark.queryQueryableIndex  avgt   25  577079.809 ± 20934.991  us/op
GroupByBenchmark.queryQueryableIndex  avgt   25  569389.209 ± 21105.720  us/op
GroupByBenchmark.queryQueryableIndex  avgt   25  563134.815 ± 14382.143  us/op


RELATIVE
GroupByBenchmark.queryQueryableIndex  avgt   25  610461.143 ± 23389.105  us/op
GroupByBenchmark.queryQueryableIndex  avgt   25  612250.288 ± 25037.477  us/op
GroupByBenchmark.queryQueryableIndex  avgt   25  609319.922 ± 20519.392  us/op
```